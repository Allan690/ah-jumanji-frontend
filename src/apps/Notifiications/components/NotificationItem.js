import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import profileAvatar from '../../../assets/img/profile-avatar.png';
import * as notificationActions from "../actions/notificationActions";


class NotificationItem extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      status: props.notification.unread
    };
    this.onClick = this.onClick.bind(this);
    this.switchStatus = this.switchStatus.bind(this);
  }

  onClick() {
    let {notification, actions} = this.props;
    let {status} = this.state;
    actions.toggleSpecificNotificationReadStatus(
      this.switchStatus(status),
      notification.id
    );

  }

  switchStatus(status) {
    switch (status) {
      case true:
        this.setState(
          {status: false}
        );
        return "mark_read";
      case false:
        this.setState(
          {status: true}
        );
        return "mark_unread";
    }
  }

  switchClass() {
    const {status} = this.state;
    switch (status) {
      case true:
        return "btn btn-light btn-sm unread";
      case false:
        return "btn btn-light btn-sm read";
    }
  }

  switchTooltip() {
    const {status} = this.state;
    switch (status) {
      case true:
        return "mark as read";
      case false:
        return "mark as unread";
    }
  }

  render() {
    const {notification,listClass} = this.props;
    const {status} = this.state;

    let default_avatar = notification.actor.profile_image || profileAvatar;
    return (
      <li className={listClass}>
        <img
          src={default_avatar} alt="avatar"
          className="followers-avatar rounded-circle" />
        <button type="button" className="btn btn-link">{notification.actor.username}</button>
        {notification.verb}
        <span className="float-right">
          <button type="button" data-toggle="tooltip" data-placement="top" title={this.switchTooltip()} className={this.switchClass()} onClick={this.onClick}>
            <i className="fas fa-check-circle" />
            {notification.status}
          </button>
        </span>
        <br />
        <i className="text-muted">
          <span className="fas fa-clock " />
          {notification.timesince}
        </i>
      </li>
    );
  }

}

NotificationItem.propTypes = {
  notification: PropTypes.object.isRequired,
  listClass: PropTypes.string.isRequired
};

function mapStateToProps(state) {
  return {
    notifications: state.Notifications
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(notificationActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(NotificationItem);
