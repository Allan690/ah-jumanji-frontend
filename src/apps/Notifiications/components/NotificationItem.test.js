import React from 'react';
import Enzyme, {shallow} from "enzyme/build";
import Adapter from "enzyme-adapter-react-16/build";
import thunk from "redux-thunk";
import TestRenderer from 'react-test-renderer';
import configureMockStore from "redux-mock-store";
import {shallowWithStore} from "../../resetPassword/tests/email_reset.test";
import NotificationItem from "./NotificationItem";
import {Provider} from "react-redux";


Enzyme.configure({adapter: new Adapter()});

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);


describe('Tests notifications item', function () {
  const InitialState = {
    Notifications: {}
  };
  const notification = {
    "id": 161,
    "unread": true,
    "emailed": false,
    "actor": {
      "user": 39,
      "username": "fgkinus2",
      "first_name": "fgkinus2",
      "last_name": "",
      "bio": "",
      "profile_photo": "",
      "country": "",
      "phone_number": "",
      "twitter_handle": "",
      "website": "",
      "created": "2019-01-10T08:37:44.560203Z",
      "modified": "2019-01-18T09:29:44.120076Z"
    },
    "verb": "Commented on the article",
    "recipient": {
      "email": "fgkinus@gmail.com",
      "username": "fgkinus2",
      "profile_image": ""
    },
    "target": null,
    "action_object": "/api/articles/title-cdnoeazysnkcqpmrgqcj-text/",
    "timestamp": "2019-01-24T17:28:21.471818Z",
    "timesince": "1Â minute"
  };
  const store = mockStore(InitialState);
  let wrapper = shallowWithStore(<NotificationItem notification={notification} listClass="list" />, store);

  it('should render correctly', function () {
    expect(wrapper.find('nav')).toBeDefined();
    expect(wrapper.find('.followers-avatar')).toBeDefined();
    expect(wrapper.find('.btn')).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should render again correctly', function () {
    const component = TestRenderer.create(
      <Provider store={mockStore({})}>
        <NotificationItem notification={notification} listClass="list" />
      </Provider>
    );


    component.unmount();
  });

  describe('<NotificationItem />', () => {
    it('should match the snapshot', () => {
      const component = shallow(<NotificationItem debug />);
      expect(component).toMatchSnapshot();
    });
  });
});
