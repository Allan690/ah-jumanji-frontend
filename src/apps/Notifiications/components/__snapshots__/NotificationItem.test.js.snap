// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collect password updates should render  1`] = `
<div
  className="row login-container"
>
  <div
    className="col-md-6 offset-md-3"
  >
    <div
      className="bg-circle-1 bg-circle"
    />
    <div
      className="bg-circle-2 bg-circle"
    />
    <div
      className="bg-circle-3 bg-circle"
    />
    <Connect(withRouter(CodeCollectionForm))
      reset_code=""
    />
  </div>
</div>
`;

exports[`ForgotPasswordComponent should render   1`] = `
<div
  className="row login-container"
>
  <div
    className="col-md-6 offset-md-3"
  >
    <div
      className="bg-circle-1 bg-circle"
    />
    <div
      className="bg-circle-2 bg-circle"
    />
    <div
      className="bg-circle-3 bg-circle"
    />
    <Connect(withRouter(EmailReset)) />
  </div>
</div>
`;

exports[`Render code collection form  should render the form 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Render reset email component should render the form 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`Tests notifications item should render correctly 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
